torch/utils/waitcounterfilelock.py

```diff
@@ -0,0 +1,37 @@
+from types import TracebackType
+from typing import Self
+
+from filelock import FileLock as base_FileLock
+
+from torch.monitor import _WaitCounter
+
+
+class WaitCounterFileLock(base_FileLock):
+    """
+    This behaves like a normal file lock.
+
+    However, it adds waitcounters for acquiring and releasing the filelock
+    as well as for the critical region within it.
+
+    pytorch.filelock.enter - While we're acquiring the filelock.
+    pytorch.filelock.region - While we're holding the filelock and doing work.
+    pytorch.filelock.exit - While we're releasing the filelock.
+    """
+
+    def __enter__(self) -> Self:
+        self.region_counter = _WaitCounter("pytorch.filelock.region").guard()
+        self.region_counter.__enter__()
+        with _WaitCounter("pytorch.filelock.enter").guard():
+            super().__enter__()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> None:
+        with _WaitCounter("pytorch.filelock.exit").guard():
+            super().__exit__(exc_type, exc_value, traceback)
+        self.region_counter.__exit__()
+        return None
```

Comment:
This is annoying and needs to be CR'ed carefully

Reply:
I think you need to test the return value of the exit calls

Reply:
I'm also not convinced you need to do it manually like this and you couldn't have just implemented this with an ExitStack or something

Reply:
re: super().__exit__ - it returns None
https://github.com/tox-dev/filelock/blob/57f488ff8fdc2193572efe102408fb63cfefe4e4/src/filelock/_api.py#L379

re, WaitCounter.__exit__() it returns nothing as well - https://www.internalfb.com/code/fbsource/[c61093c0abca]/fbcode/caffe2/torch/csrc/monitor/python_init.cpp?lines=315

So it's not clear to me what return value exit calls you want me to check.

Reply:
re: exit_stack, I played around with using it. You can use it for the WaitCounters mostly, but I couldn't figure out  a way to get it to call __enter__ and __exit__ on the parent class via super.

The other option is to implement a passthrough init class, and just make the filelock a existstack. You do have to handle passing through attrs, as well as __init__, which seemed harder than just only making some changes to enter and exit, but if you feel strongly, I'm happy to write the code that way.

Reply:
> re: super().__exit__ - it returns None
[tox-dev/filelock@57f488f/src/filelock/_api.py#L379](https://github.com/tox-dev/filelock/blob/57f488ff8fdc2193572efe102408fb63cfefe4e4/src/filelock/_api.py#L379)

> re: WaitCounter.__exit__() it returns nothing as well - [https://www.internalfb.com/code/fbsource/[c61093c0abca]/fbcode/caffe2/torch/csrc/monitor/python_init.cpp?lines=315](https://www.internalfb.com/code/fbsource/%5Bc61093c0abca%5D/fbcode/caffe2/torch/csrc/monitor/python_init.cpp?lines=315)

Include these receipts in code. It is especially important for filelock because this is third party code and can potentially change.

Reply:
Comments added :).